main_boaz.tex

```diff
@@ -0,0 +1,910 @@
+We will use naming to avoid relying on order for inputs of these tensors.
+We will now define the main concepts we use:
+
+\begin{itemize}
+
+  
+  \item An \emph{axis} is a set of the form $\{ (\text{ax} ,\text{annot},x) | x\in X \}$ where $\text{ax}$ and $\text{annot}$ are strings and $X$ is a well ordered set that can be infinite or finite. (For example, $X$ might be the integers, strings, tuples of integers or strings, or finite subsets of any of those.)   The string $\text{ax}$ is the \emph{name} of the axis, the string $\text{annot}$ is its annotation (can often be the empty string $\text{""}$). An axis with name $\text{ax}$ and empty annotation is denoted by $\name{ax}$. If there is an annotation then we use superscript, subscripts or primes to indicate it. The  set $X$ is the \emph{support} of the axis, and is denoted by $Supp(\name{name})$.  We use the notation $\name{ax} \defaxis X$ to denote that $\name{ax}$ is an axis of the form $\{ (\text{ax},\text{""},x) | x\in  X \}$.  We use the notation $\name{ax}^{annot} \defaxis X$ to denote that $\name{ax}^{\text{annot}}$ is an axis of the form $\{(\text{ax},\text{annot},x) | x\in X \}$. The support of an annotated axis $\name{ax}^{annot}$ is always a subset of the support of the axis $\name{ax}$ with an empty annotation.  If two axes have the same name and annotation then they must have the same support. If $\name{ax}^{annot}$ is an axis and $x \in Supp(X)$, then we denote by $\name{ax}[x]$ the element $(\text{ax},\text{annot},x) \in \name{ax}^{annot}$.  Our default notion for a supporting set will be the natural numbers $\mathbb{N}$ (without zero if we use one-based indexing) and so if we simply say ``''$\name{ax}$ is an ax'' then we assume that its support is the natural numbers, and hence $\name{ax}$ is the set $\{ \name{ax}[1],\name{ax}[2],\ldots \}$.\footnote{This document uses a one-based axising convention, but we can of course do everything with a zero-based axising.}
+
+
+  \item An element of an axis $\name{ax}$ is known as an \emph{index}. We think of an axis $\name{name}$ as describing a \emph{type} that inherits from the type $X$. Therefore, if (for example) $Supp(\name{ax})$ is the integers then we can use operations such as addition, subtraction, multiplication on indices.  Even if two axes have the same support, we still require explicit casting to translate between one to the other, and so (for example) if $i \in \name{batch}$ then it cannot be used as is to axis into a vector of shape $\{ \name{time} \}$. In contrast, if two axes differ only in annotation, then we can translate indices between one and the other without explicit casting. We can think of an annotated axis of the form $\name{ax}_annot$ as a parameter of type $\name{ax}$ and name \texttt{ax\_annot}. (In this sense annotation is like ``Hungarian notation'' for parameter names.) 
```

Comment:
Does the ability to implicitly convert between (ax, 1, x) and (ax, 2, x) mean that they are for all intents and purposes equivalent?

Reply:
Yes, annotations are only needed to disambiguate when you have more than one axis of the same name.

The way I think about it is that if the tensor was a python function mapping indices to real numbers, then the axis name is the **type** of the parameter, and the parameter name will be the type postdixed by the annotation.

So if A is a tensor of shape { **width** , **height** } then using python type hints we can think of it as a function defined as 

def A(width: width, height : height) : float

If B is a tensor of shape { **state**^in , **state**^out } then using python type hints we can think of it as a function called as 

def A(state_in: state , state_out : out) : float

Reply:
I think it would be good to state in a more formal way when (ax, 1, x) and (ax, 2, x) are considered equivalent and when they are considered not equivalent. The reason I think this is important is that I'm worried that these triples are doing double-duty as indices with types and as indices with names. I suspect that resolving the question of the equivalence/nonequivalence of (ax, 1, x) and (ax, 2, x) will result in a cleaner separation between index types and index names.

Reply:
Would something like the following work? (This is what I had in mind)

If you use an indexing operation A(i) were  i is an index of the form (ax,annot,x) then we do the following:

If ax^{annot} is in type of  A then  A(i) is the tensor of one smaller order obtained by restricting to this coordinate equal to i. 

If ax^{annot} is not in type A, but there is only a single axis of the form ax or ax^{annot'} in the type of A then we cast to that type.

Otherwise, we ignore this index, or decide that this is an undefined operation (depending on what convention we want to use).

Reply:
OK, so the automatic casting occurs only when indexing a tensor, that makes sense.

If I have two matrices

A ∈ ℝ^{ax^1, ax^2}
B ∈ ℝ^{ax^1, ax^2}

how would I multiply them (contracting A's ax^2 with B's ax^1)?

Or if I have a matrix and vector

A ∈ ℝ^{ax^1, ax^2}
B ∈ ℝ^{ax}

if I understand correctly,

\sum_{i ∈ ax^1} A_i B_i would multiply A^T and B
\sum_{i ∈ ax^2} A_i B_i would multiply A and B
\sum_{i ∈ ax} A_i B_i would form an order-3 tensor

which I find surprising.

Reply:
I talk about matrix multiplication in Section 2.5.3
Basically as we discussed elsewhere, if you want to apply an operator where you align two separate axes you can use a notation such as "ax1|ax2" under the dot product.

I also propose there a shorthand for matrix multiplication where you use A_{ax1} {ax2}_B (that is a subscript on the left hand of B to indicate the axis you use)

If i in ax^1 then A_i is the { ax2 } type vector obtained by A_{ax1=i}  and so the first sum is over the dot product of these vectors. I think this notation actually makes more explicit what's going on than having to keep track of rows and columns.

The second sum is similarly summing up over the dot product of  { ax2 } shaped vectors.

The third sum is simply undefined - if i is in ax, then since A has two different axes with the name "ax" then A_i is simply not defined. (It's an error to use this because it would be ambiguous.)


If we wanted AB we can also write it as sum_{i in ax} A_{ax^1[i]} B_{ax^2[i]}

Reply:
Ah, I didn't notice the additions to §2.5.3. And I see that you can use ax^1[i] to "cast" i from one annotation to another, so you could use casting to write out this operation explicitly if you wanted to.

For the second example, can you clarify -- when summing over ax^1 or ax^2, what's the resulting shape? I thought it would be a vector but your answer makes it sound like a scalar, which I would find surprising.

When summing over ax, is it not that case that the lookups A_i, where i \in ax, would fall to the third case, so A_i has shape {ax^1, ax^2}? Then the result tensor would have shape {ax, ax^1, ax^2}?

Reply:
The way I view it is the following:

If i is an index of type ax^{annot} and A is a tensor then:

If A's type contains a unique axis ax^{annot'} that can be automatically cast from ax^{annot} then A_i = A_{ax^{annot'}[i]} 

If A's type contains two such axes then the operation is undefined.

If A's type contains no such axis then it depends on the convention we want to follow with missing axes. Either we say this is undefined or ignore it and hence A_i = A.

(BTW you can imagine a more sophisticated type system than simply basic types and annotations, but I am not sure I see the use cases for this added complication.) 
---

To your question regarding the interpretation of the sum. I was confused by the capital B and missed the type, and thought it was a vector and not a matrix - sorry.  I think this example shows that it is a good idea to also introduce notation to allow the author to remind the reader of the type of particular tensors, for example by writing the type as a superscript.

So if A is { ax^1, ax^2 } type and B is { ax } type then the interpretations are as follows:

If i is in ax^1 then A_i is a { ax^2 } shaped vector  and B_i is a scalar, and so indeed you get the sum of the vectors A_i*B_i 

Analogously for i in ax^2. So one of those is the matrix A times B and the other one is A^T times B. (Judging on past experience, if I don't now go over this carefully I have 50% chance to get it wrong, which is another argument in favor of our notation...)

However, if i is in ax, then A_i is undefined since it can be cast to more than one type.

Reply:
OK, with that change, I think the rule is clear here. (I think it would be better if there were simply no automatic casting, but at least I understand the rule clearly now.)

Reply:
Of course authors can avoid all automatic casting if they don't use annotations.

We can discuss if casting needs to be automatic or explicit. I think there are three different directions:

1) Automatic casting from annotated type to base type - I think that's very helpful. If you have an operation F that you defined on (for example) some state space  then you want to be able to apply it to either a { **state**^in } tensor or a { **state**^out } tensor

2) Automatic casting between annotated types. I was thinking of this as useful in the common case when you have an operation F that takes as input a { **ax**^{in} } shape tensor and outputs an { **ax**^{out} } shape one and you want to apply it iteratively. (For example in deep neural nets that could happen all the time.) 

3) Automatic casting from base type to annotated type. This can be useful to write something like \sum_{i in **ax}} v_i w_i where v_i has shape { **ax**^1 } and v_j has shape { **ax**^2 }. It can also be thought of as a special case of (2) for the case that one of the types has an empty annotation.

A priori we could also decide that 1 is automatic but 2 and 3 are not. My tendency at the moment is to say that all 3 should be automatic,  because I think it will make the notation less cumbersome and in most cases authors intentions will be clear. Authors can always write down explicit casting if they want to.  I'd suggest that if you want to explicitly cast from { **ax**^{annot} } to  { **ax**^{annot'} } then you can just use annot -> annot'  (no need to write the **ax** as well)

My high level objectives are that: 

1) Our notation should make it easy for a reader to look at an expression and determine its type. (And the author should have non intrusive ways to remind the reader of it when expressions are complicated.) 

2) Authors should be able to express their main ideas clearly, with minimal extraneous uses of primes, extra variables, and casting, unless those are needed to disambiguate.